README.md
# Tab Kit - Mouse Gestures
Provide some mouse gesture on tabs and other stuff

## Features
- Navigate to last page by clicking LMB while holding RMB (default: on)
- Switch tabs by scroll wheel when hover on tab bar
- Switch tabs by scroll wheel when holding right mouse button
- Switch to a tab when pointer is hovered on it
- Options can be changed inside Addons Manager

## Install
https://addons.mozilla.org/en-US/firefox/addon/tab-kit-mouse-gestures/

## Changes
See https://github.com/tabkit/mouse-gestures/blob/master/CHANGELOG.md for what's changed

## Tab Kit Series Extensions
- [Tab Kit 2nd Edition](https://github.com/tabkit/tabkit2)
- [Tab Kit - Tab Highlighter](https://github.com/tabkit/tab-highlighter)
- [Tab Kit - Mouse Gestures](https://github.com/tabkit/mouse-gestures)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project does NOT adhere to [Semantic Versioning](http://semver.org/) YET (Will do so after rewrite).


## [Unreleased][unreleased] - YYYY-MM-DD

### Added
- Nothing

### Changed
- Nothing

### Fixed
- Nothing


## [0.4.0][0.4.0] - 2019-06-23

### Added
- New option for enable/disable "Scroll Wheel Tab Switching over the edge"  
  (Issue: https://github.com/tabkit/mouse-gestures/issues/6)


## [0.3.3][0.3.3] - 2018-02-22

### Changed
- Target Pale Moon instead of Firefox


## [0.3.3][0.3.3] - 2018-02-22

### Changed
- Target Pale Moon instead of Firefox


## [0.3.2.1][0.3.2.1] - 2017-03-04

### Fixed
- Fix switch tab by scrolling on Tab Bar in FF 52

## [0.3.1][0.3.1] - 2016-06-19

### Added
- New option for custom delay of "Switch to a tab when pointer is hovered on it"


## [0.3.0][0.3.0] - 2016-06-19

### Added
- New option for closing tab with LMB click with Ctrl/Cmd pressed

### Fixed
- Some mouse gestures did not work properly when cursor is in tab container (tab bar)


## [0.2.0][0.2.0] - 2015-04-17

### Added
- Switch tabs by scroll wheel when holding left mouse button


## [0.1.2][0.1.2] - 2013-07-28

### Fixed
- Re-run initialization code after failure


## [0.1.1][0.1.1] - 2013-05-17

### Fixed
- Delay is working for "Switch to a tab when cursor is hovered on it"
- Context Menu no longer show up when using "Switch tabs by scroll wheel when holding right mouse button"


## 0.1.0 - 2013-05-01

### Added
- Navigate to last page by clicking LMB while holding RMB (default: on)
- Switch tabs by scroll wheel when hover on tab bar
- Switch tabs by scroll wheel when holding right mouse button
- Switch to a tab when cursor is hovered on it


[unreleased]:   https://github.com/tabkit/mouse-gestures/compare/v0.4.0...HEAD
[0.4.0]:        https://github.com/tabkit/mouse-gestures/compare/v0.3.3...v0.4.0
[0.3.3]:        https://github.com/tabkit/mouse-gestures/compare/v0.3.2.1...v0.3.3
[0.3.2.1]:      https://github.com/tabkit/mouse-gestures/compare/v0.3.1...v0.3.2.1
[0.3.1]:        https://github.com/tabkit/mouse-gestures/compare/v0.3.0...v0.3.1
[0.3.0]:        https://github.com/tabkit/mouse-gestures/compare/v0.2.0...v0.3.0
[0.2.0]:        https://github.com/tabkit/mouse-gestures/compare/v0.1.2...v0.2.0
[0.1.2]:        https://github.com/tabkit/mouse-gestures/compare/v0.1.1...v0.1.2
[0.1.1]:        https://github.com/tabkit/mouse-gestures/compare/v0.1.0...v0.1.1


bootstrap.js
"use strict";

//==================== Modulize JS Scripts START ====================//

/* Includes a javascript file with loadSubScript
*
* @param src (String)
* The url of a javascript file to include.
*/
(function(global) {
  global.include = function include(src) {
    const o = {}
    Components.utils.import("resource://gre/modules/Services.jsm", o)
    const uri = o.Services.io.newURI(
      src, null, o.Services.io.newURI(__SCRIPT_URI_SPEC__, null, null))
    o.Services.scriptloader.loadSubScript(uri.spec, global)
  }
})(this);

/* Imports a commonjs style javascript file with loadSubScript
*
* @param src (String)
* The url of a javascript file.
*/
(function(global) {
  const modules = {}
  global.require = function require(src) {
    if (modules[src]) return modules[src]
    const scope = {require: global.require, exports: {}}
    const tools = {}
    Components.utils.import("resource://gre/modules/Services.jsm", tools)
    const baseURI = tools.Services.io.newURI(__SCRIPT_URI_SPEC__, null, null)
    try {
      const uri = tools.Services.io.newURI(
          "packages/" + src + ".js", null, baseURI)
      tools.Services.scriptloader.loadSubScript(uri.spec, scope)
    } catch (e) {
      const uri = tools.Services.io.newURI(src, null, baseURI)
      tools.Services.scriptloader.loadSubScript(uri.spec, scope)
    }
    return modules[src] = scope.exports
  }
})(this)

//==================== Modulize JS Scripts END ====================//

// Constants
const { classes: Cc, Constructor: CC, interfaces: Ci, utils: Cu,
        results: Cr, manager: Cm } = Components


// Import any Firefox modules here
Cu.import("resource://gre/modules/Services.jsm")
Cu.import("resource://gre/modules/AddonManager.jsm")


// Import any custom JS files
// But they are still part of this addon
// Just like code separated into different files
include("includes/prefs.js")
include("includes/init-listeners.js")
include("includes/mouse-history-gesture.js")


// Require any CommonJS style files
const utils = require("utils")
const {unload} = require("unload")
const {log, debug, dump} = require("console")
const prefUtils = require("pref-utils")
const sessionStore = require("session-store")


//==================== Code to run with DOM Window START ====================//
// Origin Code from http://www.oxymoronical.com/blog/2011/01/Playing-with-windows-in-restartless-bootstrapped-extensions
const WindowListener = {
  setupBrowserUI: function(window) {
    const document = window.document

    // Take any steps to add UI or anything to the browser window
    // document.getElementById() etc. will work here



    // Run all the listeners (after a timeout)
    runPostInitListeners(window)
  },

  tearDownBrowserUI: function(window) {
    const document = window.document

    // Take any steps to remove UI or anything from the browser window
    // document.getElementById() etc. will work here
  },

  // nsIWindowMediatorListener functions
  onOpenWindow: function(xulWindow) {
    // A new window has opened
    const domWindow = xulWindow.QueryInterface(Ci.nsIInterfaceRequestor)
                             .getInterface(Ci.nsIDOMWindow)

    // Wait for it to finish loading
    domWindow.addEventListener("load", function listener() {
      domWindow.removeEventListener("load", listener, false)

      // If this is a browser window then setup its UI
      if (domWindow.document.documentElement.getAttribute("windowtype") === "navigator:browser")
        WindowListener.setupBrowserUI(domWindow)
    }, false)
  },

  onCloseWindow: function(xulWindow) {
  },

  onWindowTitleChange: function(xulWindow, newTitle) {
  }
}
//==================== Code to run with DOM Window END ====================//



//==================== Bootstrap addon methods START ====================//
function install(data, reasonCode) {
  //empty
}

function uninstall(data, reasonCode) {
  //empty
}

function startup(data, reasonCode) {
  // Always set the default prefs as they disappear on restart
  prefUtils.setDefaultPrefs()

  // === Register the preference listeners ===
  // Otherwise the preference attribute mapping does not work
  prefUtils.registerPrefObserver(localPrefs)

  // === Register methods that should be run after most extensions ===
  // Since some event listeners can only be added at that time
  // Only add this once since it is run for every new window
  postInitListeners.push(initMouseGestures)


  const wm = Cc["@mozilla.org/appshell/window-mediator;1"].
           getService(Ci.nsIWindowMediator)

  // Get the list of browser windows already open
  const windows = wm.getEnumerator("navigator:browser")
  while (windows.hasMoreElements()) {
    const domWindow = windows.getNext().QueryInterface(Ci.nsIDOMWindow)

    WindowListener.setupBrowserUI(domWindow)
  }

  // Wait for any new browser windows to open
  wm.addListener(WindowListener)
}

function shutdown(data, reasonCode) {
  // When the application is shutting down we normally don't have to clean
  // up any UI changes made
  if (reasonCode === APP_SHUTDOWN)
    return


  // Run all the unload callbacks on shutdown
  unload()


  const wm = Cc["@mozilla.org/appshell/window-mediator;1"].
           getService(Ci.nsIWindowMediator)

  // Get the list of browser windows already open
  const windows = wm.getEnumerator("navigator:browser")
  while (windows.hasMoreElements()) {
    const domWindow = windows.getNext().QueryInterface(Ci.nsIDOMWindow)

    WindowListener.tearDownBrowserUI(domWindow)
  }

  // Stop listening for any new browser windows to open
  wm.removeListener(WindowListener)
}
//==================== Bootstrap addon methods END ====================//


includes/prefs.js
/*
 * Stores the required constants and hardcoded default preferences
 * Also declare global and local preference service
 */

const PREF_BRANCH = "extensions.tabkit.mouse-gestures."
// Default pref values
const PREFS = {
  tabWheelSwitchHover: false,
  tabWheelSwitchRMB: false,
  tabWheelSwitchLMB: false,
  tabWheelSwitchOverEdges: true,

  switchTabsOnHover: false,
  switch_tab_on_hover_delay_length_in_millisecond: 200,

  lmbRmbBackForward: true,
  closeTabWithCtrlLMB: false
}

// Make sure we can use gPrefService from now on (even if this isn't a browser window!)
if (typeof globalPrefs == "undefined" || !globalPrefs) {
  const globalPrefs = Cc["@mozilla.org/preferences-service;1"].
           getService(Ci.nsIPrefBranch)
}

const localPrefs = Cc["@mozilla.org/preferences-service;1"]
                 .getService(Ci.nsIPrefService)
                 .getBranch(PREF_BRANCH)


includes/mouse-history-gesture.js
// Handlers for all windows
const _handlers = []


// Run after each window start
function initMouseGestures (window) {
  const handlerIndex = _handlers.push(new MouseGesturesHandler(window)) - 1
  // Might not be necessary, but whatever
  unload(function() { handler[handlerIndex] = null })
}

const MouseGesturesHandler = function(window) {
  this._window = window


  const onMouseUpGestureBind = this.onMouseUpGesture.bind(this)
  window.gBrowser.addEventListener("mouseup", onMouseUpGestureBind, true)
  unload(function() { window.gBrowser.removeEventListener("mouseup", onMouseUpGestureBind, true) })
  window.gBrowser.tabContainer.addEventListener("mouseup", onMouseUpGestureBind, true)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("mouseup", onMouseUpGestureBind, true) })

  const onMouseDownGestureBind = this.onMouseDownGesture.bind(this)
  window.gBrowser.addEventListener("mousedown", onMouseDownGestureBind, true)
  unload(function() { window.gBrowser.removeEventListener("mousedown", onMouseDownGestureBind, true) })
  window.gBrowser.tabContainer.addEventListener("mousedown", onMouseDownGestureBind, true)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("mousedown", onMouseDownGestureBind, true) })

  const onContextMenuGestureBind = this.onContextMenuGesture.bind(this)
  window.gBrowser.addEventListener("contextmenu", onContextMenuGestureBind, true)
  unload(function() { window.gBrowser.removeEventListener("contextmenu", onContextMenuGestureBind, true) })
  window.gBrowser.tabContainer.addEventListener("contextmenu", onContextMenuGestureBind, true)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("contextmenu", onContextMenuGestureBind, true) })

  const onMouseDragGestureBind = this.onMouseDragGesture.bind(this)
  window.gBrowser.addEventListener("dragstart", onMouseDragGestureBind, true)
  unload(function() { window.gBrowser.removeEventListener("dragstart", onMouseDragGestureBind, true) })
  window.gBrowser.tabContainer.addEventListener("dragstart", onMouseDragGestureBind, true)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("dragstart", onMouseDragGestureBind, true) })

  const onMouseOutGestureBind = this.onMouseOutGesture.bind(this)
  window.gBrowser.addEventListener("mouseout", onMouseOutGestureBind, true)
  unload(function() { window.gBrowser.removeEventListener("mouseout", onMouseOutGestureBind, true) })


  const onTabWheelGestureBind = this.onTabWheelGesture.bind(this)
  window.gBrowser.tabContainer.addEventListener("wheel", onTabWheelGestureBind, true)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("wheel", onTabWheelGestureBind, true) })

  const onLMBWheelGestureBind = this.onLMBWheelGesture.bind(this)
  window.gBrowser.addEventListener("wheel", onLMBWheelGestureBind, true)
  unload(function() { window.gBrowser.removeEventListener("wheel", onLMBWheelGestureBind, true) })

  const onRMBWheelGestureBind = this.onRMBWheelGesture.bind(this)
  window.gBrowser.addEventListener("wheel", onRMBWheelGestureBind, true)
  unload(function() { window.gBrowser.removeEventListener("wheel", onRMBWheelGestureBind, true) })

  const resetMouseScrollWrapCounterBind = this.resetMouseScrollWrapCounter.bind(this)
  window.gBrowser.tabContainer.addEventListener("TabSelect", resetMouseScrollWrapCounterBind, false)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("TabSelect", resetMouseScrollWrapCounterBind, false) })


  const onTabHoverGestureBind = this.onTabHoverGesture.bind(this)
  window.gBrowser.tabContainer.addEventListener("mouseover", onTabHoverGestureBind, false)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("mouseover", onTabHoverGestureBind, false) })

  const cancelTabHoverGestureBind = this.cancelTabHoverGesture.bind(this)
  window.gBrowser.tabContainer.addEventListener("mouseout", cancelTabHoverGestureBind, false)
  unload(function() { window.gBrowser.tabContainer.removeEventListener("mouseout", cancelTabHoverGestureBind, false) })
}

MouseGesturesHandler.prototype = {
  _window: null,
  _mousedown: [false, undefined, false],

  _shouldPreventContext: false,
  _mouseScrollWrapCounter: 0,

  _hoverTab: null,
  _hoverTimer: null,


  onMouseUpGesture: function (event) {
    if (!event.isTrusted)
      return

    // Ignore if splitter is being dragged
    const splitter = this._window.document.getElementById("tabkit-splitter")
    if (splitter && splitter.getAttribute("state") === "dragging")
      return

    const btn = event.button
    if (this._mousedown[btn])
      this._mousedown[btn] = false
    else if (btn !== 1)
      event.preventDefault() // We've probably just done a rocker gesture
  },

  onMouseDownGesture: function (event) {
    if (!event.isTrusted)
      return

    // Ignore if splitter is being dragged
    const splitter = this._window.document.getElementById("tabkit-splitter")
    if (splitter && splitter.getAttribute("state") === "dragging")
      return

    //0 = LMB, 2 = RMB
    const btn = event.button
    let opp  //opposite button
    if (btn === 0) {
      opp = 2
    }
    else if (btn === 2) {
      opp = 0
    }

    if (this._mousedown[opp] && prefUtils.getPref("lmbRmbBackForward")) {
      if (btn === 0)
        this._window.BrowserBack()
      else
        this._window.BrowserForward()

      this._shouldPreventContext = true
      // Since the Firefox loses mouseup events during the page load (http://forums.mozillazine.org/viewtopic.php?p=33605#33605)
      this._mousedown[opp] = false

      event.preventDefault()
      event.stopPropagation()
    }
    else {
      this._mousedown[btn] = true
    }

    const tab = event.target
    // Cmd key represented by #metaKey
    // http://www.w3.org/TR/DOM-Level-3-Events/
    if (btn === 0 && (event.ctrlKey || event.metaKey)) {
      if (tab && tab.localName === "tab" && prefUtils.getPref("closeTabWithCtrlLMB")) {
        this._window.gBrowser.removeTab(tab)
      }
    }
  },


  // Prevent context menu to be displayed if
  // _shouldPreventContext is true
  onContextMenuGesture: function (event) {
    if (!event.isTrusted || !this._shouldPreventContext)
      return

    event.preventDefault()
    event.stopPropagation()

    this._shouldPreventContext = false
  },

  onMouseDragGesture: function (event) {
    if (!event.isTrusted)
      return

    this._mousedown[0] = this._mousedown[2] = false
  },

  onMouseOutGesture: function (event) {
    if (!event.isTrusted || event.target !== event.currentTarget) // n.b. this refers to gBrowser, not tabkit!
      return

    //comment it for fixing Issue 23(RMBWheelGesture not work after one scroll)
    // this._mousedown[2] = false
    this._mousedown[0] = false
  },


  onTabWheelGesture: function (event) {
    if (!event.isTrusted) {
      return
    }

    const name = event.originalTarget.localName

    if (name === "scrollbar" || name === "scrollbarbutton" || name === "slider" || name === "thumb") {
      return
    }

    if (prefUtils.getPref("tabWheelSwitchHover")) {
      this.scrollwheelTabSwitch(event)
    }
  },

  onLMBWheelGesture: function(event) {
    if (!event.isTrusted || !this._mousedown[0] || !prefUtils.getPref("tabWheelSwitchLMB"))
      return

    this.scrollwheelTabSwitch(event)
    if (event.change !== 0) {
      this._shouldPreventContext = true
    }
  },

  onRMBWheelGesture: function(event) {
    if (!event.isTrusted || !this._mousedown[2] || !prefUtils.getPref("tabWheelSwitchRMB"))
      return

    this.scrollwheelTabSwitch(event)
    if (event.change !== 0) {
      this._shouldPreventContext = true
    }
  },

  // To be called by onTabWheelGesture & onRMBWheelGesture
  // Not to be binded with any event
  scrollwheelTabSwitch: function (event) {
    const change = event.deltaY

    // Scrolling up
    if (change > 0) {
      let lastTab = this._window.gBrowser.tabs[this._window.gBrowser.tabs.length - 1]

      // Skip hidden tabs
      while (lastTab.hidden && lastTab.previousSibling) {
        lastTab = lastTab.previousSibling
      }

      // Switch to next tab, but requiring 3 wheelscrolls to wrap around
      if (this._window.gBrowser.tabContainer.selectedIndex < lastTab._tPos || this._mouseScrollWrapCounter >= 2) {
        this._window.gBrowser.tabContainer.advanceSelectedTab(1, true)
        // Note: _mouseScrollWrapCounter is reset whenever a tab is selected
      }
      // Only increment the counter when preference checked
      else if (prefUtils.getPref("tabWheelSwitchOverEdges")) {
        this._mouseScrollWrapCounter++
      }
    }
    // Scrolling down
    else if (change < 0) {
      let firstTab = this._window.gBrowser.tabs[0]

      // Skip hidden tabs
      while (firstTab.hidden && firstTab.nextSibling) {
        firstTab = firstTab.nextSibling
      }

      // Switch to previous tab, but requiring 3 wheelscrolls to wrap around
      if (this._window.gBrowser.tabContainer.selectedIndex > firstTab._tPos || this._mouseScrollWrapCounter >= 2) {
        this._window.gBrowser.tabContainer.advanceSelectedTab(-1, true)
        // Note: _mouseScrollWrapCounter is reset whenever a tab is selected
      }
      // Only increment the counter when preference checked
      else if (prefUtils.getPref("tabWheelSwitchOverEdges")) {
        this._mouseScrollWrapCounter++
      }
    }

    event.preventDefault()
    event.stopPropagation()
  },

  resetMouseScrollWrapCounter: function (event) {
    this._mouseScrollWrapCounter = 0
  },


  onTabHoverGesture: function (event) {
    if (!event.isTrusted || event.target.localName !== "tab" || !prefUtils.getPref("switchTabsOnHover")) {
      return
    }

    this._hoverTab = event.target

    // Minimum delay, might make it a config later
    const DEFAULT_VALUE = 200
    let value_from_pref = prefUtils.getPref("switch_tab_on_hover_delay_length_in_millisecond")
    let waitTime = (
      typeof value_from_pref !== "number" ||
      isNaN(value_from_pref) ||
      value_from_pref < 0
    ) ? DEFAULT_VALUE : value_from_pref
    // See whether the tab hovering on is just next to the current tab
    const isNeighbourTab = (Math.abs(this._hoverTab._tPos - this._window.gBrowser.selectedTab._tPos) === 1)
    // if hovering neighbour tab, use a longer delay in case it is an accident hover (especially in vertical mode)
    if (isNeighbourTab) {
      waitTime = waitTime * 2
    }

    this._window.clearTimeout(this._hoverTimer)

    this._hoverTimer = this._window.setTimeout(function () {
      this._window.gBrowser.selectedTab = this._hoverTab
    }.bind(this), waitTime)
  },

  cancelTabHoverGesture: function (event) {
    if (!event.isTrusted || event.target !== this._hoverTab)
      return

    this._window.clearTimeout(this._hoverTimer)
  }
}
